lib/pbench/agent/tool_data_sink.py

```diff
@@ -186,6 +190,408 @@ def terminate(self):
         return 1
 
 
+class PCPPmlogger(BaseCollector):
+    """ Log pcp metric data """
+
+    def __init__(self, *args, **kwargs):
+        """
+        input: takes list of hosts and tools
+        what init function does:
+        1. checks whether pmcd is active amongst the hosts mentions in the list.
+        2. reads the pcp-mappig file.
+        3. builds the pmlogger configuration based on the tools based on the list using pmlogconf tool.
+        4. builds the control file required by pmlogger_check tool.
+        for more info on pmlogconf and pmlogger_check check respective man pages.
+        """
+        super().__init__(*args, **kwargs)
+        # save external state
+        self.temp_hosts = list(self.host_tools_dict.keys())
+        # initialize internal state
+        self.hosts = []
+        self.pmlog_config_files = []
+        self.mapping = dict()
+        self.control_file = os.path.join(
+            self.benchmark_run_dir, ".temp", "pcppmlogger.d"
+        )
+        self.log_dir = os.path.join(self.benchmark_run_dir, "results")
+        # call init functions
+        self.check_connection(self.temp_hosts)
+        # WILL LIKELY BE AN ISSUE HERE
+        try:
+            inst_dir = PbenchAgentConfig(
+                os.environ["_PBENCH_AGENT_CONFIG"]
+            ).pbench_install_dir
+        except Exception as exc:
+            self.logger.error(
+                "Unexpected error encountered logging pbench agent configuration: '%s'",
+                exc,
+            )
+        self.read_json(os.path.join(inst_dir, "util-scripts", "pcp-mapping.json"))
+        self.build_pmlog_configs(self.hosts)
+        self.build_control_file(self.hosts)
+
+    def start_logger(self):
+        """ starting logging process """
+        logging.basicConfig(
+            filename="pmlogger.log",
+            filemode="w",
+            format="%(asctime)s,%(msecs)d %(name)s %(levelname)s %(message)s",
+            datefmt="%H:%M:%S",
+            level=logging.DEBUG,
+        )
+
+        self.logger = logging.getLogger("pmlogger")
+
+    def read_json(self, json_fn):
+        """ read json file """
+        try:
+            self.logger.info("Reading the pcp mapping file")
+            self.logger.debug("Reading file:%s", json_fn)
+            with open(os.path.join(json_fn)) as json_file:
+                self.mapping = json.loads(json_file.read())
+        except Exception as e:
+            self.logger.critical(
+                "Not able to load the mapping file\n:%s", e, exc_info=True
+            )
+
+    def start_logging(self):
+        """ invoke pmlogger and start logging using pmlogger_check -c """
+        # code to start logging
+        self.logger.info("starting logger")
+
+        command = PCP.pmGetConfig("PCP_BINADM_DIR") + "/pmlogger_check -c {}".format(
```

Comment:
As disked verbally today, let's work on containerizing the `pmlogger` much like the way we handle Prometheus.